Fix issue where static pods never receive kubernetes labels

joestringer opened the pull request:
When a pod's labels are unable to be resolved in kube-apiserver, start
up a controller which will regularly attempt resolving the pod labels.
Eventually the apiserver should become aware of the pod and the labels
will be resolved, at which point this new code will update the
labels/identity and stop the controller.

We need this because apiserver is not sending notifications to Cilium
about static pods when it learns about them, so we would otherwise
never learn the labels of such pods.

Related: #9255
Fixes: #8559

I manually validated this by:
* Deploying this version of Cilium
* Configuring kubelet to support static pods (restart required)
* Disabled apiserver
* Created a static pod yaml in the configured kubelet directory
* Restarted apiserver after a couple of seconds
* Observed that the new controller was running
* After some time, observed that the new controller was no longer running, and the static pod had all of the labels from kubernetes.

The above manual test validation formed the basis for the script added in the second patch.

I also briefly considered instead supporting mounting the static pod directory into the Cilium container and having cilium-agent use these files as a potential second source of truth. This would be operationally more complex (requiring configuration of the directory at the very least), and would split the source of truth from just apiserver to two locations - files and apiserver. Furthermore the pod yaml doesn't specify the full pod name (which is derived from the hostname) and may not provide the namespace, so we would have to reproduce the logic for deriving these and the cluster in that logic (perhaps not a big deal). If it is strongly important to reduce the time period in which static pods may be represented with the "init" identity then we may still want to investigate this approach instead. That said we can also tweak the controller run period in this PR to speed things up, though it is still fundamentally dependent on the apiserver being notified about the static pod.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9257)
<!-- Reviewable:end -->

joestringer committed:
daemon: Start controller when pod labels resolution fails

When a pod's labels are unable to be resolved in kube-apiserver, start
up a controller which will regularly attempt resolving the pod labels.
Eventually the apiserver should become aware of the pod and the labels
will be resolved, at which point this new code will update the
labels/identity and stop the controller.

Signed-off-by: Joe Stringer <joe@cilium.io>

ianvernon commented on daemon/endpoint.go:
```diff
@@ -246,6 +249,47 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
+		if !k8sLabelsConfigured {
+			done := make(chan struct{})
+
+			controllerName := fmt.Sprintf("resolve-labels-%d", ep.GetID())
```
Will the endpoint ID be non-zero here? The call to `AddEndpoint` on the endpointmanager is performed immediately after this logic, and the endpointmanager is the entity which assigns an ID to an endpoint.

joestringer replied:
Argh, good point. I can construct it based on the pod namespace/name instead.

aanm commented on daemon/endpoint.go:
```diff
@@ -246,6 +249,47 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
+		if k8sLabelsConfigured {
+			// If there are labels, but no pod namespace then it's
+			// likely that there are no k8s labels at all. Resolve.
+			_, k8sNamespaceSet := addLabels[k8sConst.PodNamespaceLabel]
```
You can simply do `_, k8sLabelsConfigured := addLabels[k8sConst.PodNamespaceLabel]` in line 257

gandro commented on daemon/endpoint.go:
```diff
@@ -246,6 +249,47 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
+	// static pod's labels. In this case, start a controller to attempt to
+	// resolve the labels.
+	if ep.K8sNamespaceAndPodNameIsSet() && k8s.IsEnabled() {
+		k8sLabelsConfigured := len(addLabels) != 0
```
I'm not super familiar with this code, so I might be missing something. But will `k8sLabelsConfigured` not always be true due to the assignment a few lines above?
https://github.com/cilium/cilium/blob/9450af1d2142c8c841d42aac8d73e97b869c12e2/daemon/endpoint.go#L243-L250

I believe the fallback to the `reserved:init` identity needs to be moved to after the static pod case.

joestringer replied:
Ah, good point, I was debating whether to place the labels inside that check or outside. I could either drop the Len 0 check here or move the init identity configuration bit later.

aanm reviewed (commented):
I don't understand why these changes are required. I've test out cilium without this patch and the endpoint got the labels from kube-apiserver as soon it showed up there.

joestringer committed:
test: Add a standalone test for validating static pod labels

This test works in a microk8s environment, and is quite
environment-specific which makes it difficult to integrate properly into
the main testsuite. It is intended to be run manually for basic
validation of the issue fixed in PR #9257.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm reviewed (approved):
I've test this out with container-runtime=disabled and it works pretty well :100:
